### ISSUE TITLE
Gradient accumulations not working as expected

### ISSUE THREAD

**Issue by angelrodriguez5** [eriklindernoren/PyTorch-YOLOv3#325]
In train.py:

            loss, outputs = model(imgs, targets)
            loss.backward()

            if batches_done % opt.gradient_accumulations:
                # Accumulates gradient before each step
                optimizer.step()
                optimizer.zero_grad()

When I set gradient_accumulations to be N, I expect that the optimizer will only be called  once every N batches, instead this piece of code calls the optimizer every batch except the Nth. Not only that but if you set gradient_accumulations to 1, which would mean "pass the optimizer every batch" , the optimizer never gets called because anything % 1 == 0.

SOLUTION:
Just reverse the if condition:

            if batches_done % opt.gradient_accumulations == 0:

I'm not sure if this is an error or if I'm not understanding something correctly. Anyhow, thanks for an amazing work developing this program.

**Comment by ujsyehao** [comment]
@angelrodriguez5  batched_done is sum of training batch, ex: 1, 2, 3, 4, 5,
so if you set opt.gradient_accumulations = 2, then:
1 % 2 = 1
2 % 2 = 0
3 % 2 = 1

**Comment by angelrodriguez5** [comment]
@ujsyehao that's what I mean, it only works fine for opt.gradient_accumulations = 2.

Imagine opt.gradient_accumulations = 1
1 % 1 = 0
2 % 1 = 0
3 % 1 = 0
The optimizer never gets called. If I set that parameter to 1, I would expect to accumulate the gradient just once before optimizing, that is, optimize in every batch.

And for numbers greater than 2, imagine opt.gradient_accumulations = N:

1 % N = 1 (True)
2 % N = 2 (True)
...
N % N = 0 (False)

Then the optimizer is called every batch, except on multiples of N, so the gradient is NOT accumulated N times before optimizing.

**Comment by pandalgx** [comment]
@angelrodriguez5 I also find this problem, and I think it should be changed to `if not batches_done % opt.gradient_accumulations:`

**Comment by bisonliao** [comment]
I do think it is a bug.

**Comment by Flova** [comment]
The has been fixed
